### PR TITLE
Kmean triclustering

### PR DESCRIPTION
**Pull request by rogerkuou**
Hi @fnattino,  I generalized kmeans for Nd clustering as we discussed, and added new tests. And I also added a warning if there are multiple max values for Silhouette measure.
Would you like to take a look?

### CODE REVIEW

File: cgc/kmeans.py [hunk 1]
```diff
@@ -56,33 +49,42 @@ def __init__(self,
...
-        self.n_col_clusters = n_col_clusters
+        self.clusters = clusters
+        self.nclusters = nclusters
         self.kmean_max_iter = kmean_max_iter
         self.output_filename = output_filename
 
-        nonempty_row_cl = len(np.unique(self.row_clusters))
-        nonempty_col_cl = len(np.unique(self.col_clusters))
-        max_k = nonempty_row_cl * nonempty_col_cl
+        max_k = np.cumprod(self.nclusters)[-1]
```
Comment (fnattino) [inline cgc/kmeans.py:57]: Maybe simpler:
```suggestion
        max_k = np.prod(self.nclusters)
```

File: cgc/kmeans.py [hunk 2]
```diff
@@ -56,33 +49,42 @@ def __init__(self,
...
-        # Number of row/col clusters should be smaller than the max ID
-        # Since ID starts from 0
-        if not max(self.row_clusters) < self.n_row_clusters:
-            raise ValueError("row_clusters include labels >= n_row_clusters")
-        if not max(self.col_clusters) < self.n_col_clusters:
-            raise ValueError("col_clusters include labels >= n_col_clusters")
+        # Check if Z matches the clusters
+        if Z.ndim != len(clusters):
+            raise ValueError("The number of dimensions of Z is not equal to "
+                             "the number of dimensions of the clusters: "
```
Comment (fnattino) [inline cgc/kmeans.py:73]: Perhaps:
```suggestion
                             "the number of labels provided: "
```

File: cgc/kmeans.py [hunk 3]
```diff
@@ -56,33 +49,42 @@ def __init__(self,
...
+            raise ValueError("The shape of Z does not match the shape of the "
+                             "clusters: {} != {}".format(
+                                 Z.shape, tuple(len(cl) for cl in clusters)))
+
+        # The max label per cluster should be smaller than the number of
+        # clusters. Label starts from 0.
+        for cl, ncl, id in zip(clusters, nclusters, range(Z.ndim)):
+            if not max(cl) < ncl:
+                raise ValueError(
+                    "One cluster includes labels >= number of clusters. "
```
Comment (fnattino) [inline cgc/kmeans.py:85]:
```suggestion
                    "One label array includes elements >= number of clusters. "
```

File: tests/test_kmeans.py [hunk 4]
```diff
@@ -1,17 +1,19 @@
+from cgc.results import Results
 import unittest
+from unittest.runner import TextTestResult
```
Comment (fnattino) [inline tests/test_kmeans.py:3]: Mmh, are `Results` and `TextTestResult` needed here?

File: cgc/kmeans.py [hunk 5]
```diff
@@ -112,62 +114,84 @@ def compute(self):
...
                                             silhouette_avg)
-            kmeans_cc_list.append(kmeans_cc)
+            kmean_cluster_list.append(kmean_cluster)
         idx_k = np.argmax(silhouette_avg_list)
+        if np.sum(silhouette_avg_list == silhouette_avg_list[idx_k]) > 1:
+            idx_k_list = np.argwhere(
+                silhouette_avg_list == silhouette_avg_list[idx_k]).reshape(
+                    -1).tolist()
+            logger.warning(
+                "Multiple potential k found: {}, picking the smallest one: {}".
```
Comment (fnattino) [inline cgc/kmeans.py:134]: Maybe a bit more explicit:
```suggestion
                "Multiple k values with the same silhouette score: {},"
                "picking the smallest one: {}".
```

**Review by fnattino — COMMENTED** [review]
Great job @rogerkuou , looks very nice! I have added few suggestions, then go ahead and merge for my side!

**Comment by rogerkuou** [comment]
Thx @fnattino. I updated according to your comments. I will merge it.

**Pull request merged**